python/lsst/daf/butler/butler.py

```diff
@@ -200,23 +200,74 @@ def transaction(self):
+        method accepts those arguments and always returns a true `DatasetType`
+        and a `DataId` or `dict`.
+
+        Standardization of `dict` vs `DataId` is best handled by passing the
+        returned ``dataId`` (and ``kwds``) to `Registry` APIs, which are
+        generally similarly flexible.
+
+        Parameters
+        ----------
+        datasetRefOrType : `DatasetRef`, `DatasetType`, or `str`
+            When `DatasetRef` the `dataId` should be `None`.
+            Otherwise the `DatasetType` or name thereof.
+        dataId : `dict` or `DataId`
+            A `dict` of `Dimension` link name, value pairs that label the
+            `DatasetRef` within a Collection. When `None`, a `DatasetRef`
+            should be provided as the second argument.
+        kwds : `dict`
+            Additional keyword arguments used to construct or augment the
+            `DataId`.
+
+        Returns
+        -------
+        datasetType : `DatasetType`
+            A `DatasetType` instance extracted from ``datasetRefOrType``.
+        dataId : `dict` or `DataId`, optional
+            Argument that can be used (along with ``kwds``) to construct a
+            `DataId`.
+        """
+        if isinstance(datasetRefOrType, DatasetRef):
+            if dataId is not None or kwds:
+                raise ValueError("DatasetRef given, cannot use dataId as well")
+            datasetType = datasetRefOrType.datasetType
+            dataId = datasetRefOrType.dataId
+        else:
+            # Don't check whether DataId is provided, because Registry APIs
+            # can usually construct a better error message when it wasn't.
+            if isinstance(datasetRefOrType, DatasetType):
+                datasetType = datasetRefOrType
+            else:
+                datasetType = self.registry.getDatasetType(datasetRefOrType)
+        return datasetType, dataId
+
     @transactional
-    def put(self, obj, datasetRefOrType, dataId=None, producer=None):
+    def put(self, obj, datasetRefOrType, dataId=None, producer=None, **kwds):
         """Store and register a dataset.
 
         Parameters
         ----------
         obj : `object`
             The dataset.
-        datasetRefOrType : `DatasetRef`, `DatasetType` instance or `str`
+        datasetRefOrType : `DatasetRef`, `DatasetType`, or `str`
             When `DatasetRef` the `dataId` should be `None`.
             Otherwise the `DatasetType` or name thereof.
-        dataId : `dict`, optional
-            An identifier with `DataUnit` names and values.
-            When `None` a `DatasetRef` should be supplied as the second
-            argument.
+        dataId : `dict` or `DataId`
+            A `dict` of `Dimension` link name, value pairs that label the
+            `DatasetRef` within a Collection. When `None`, a `DatasetRef`
+            should be provided as the second argument.
         producer : `Quantum`, optional
             The producer.
+        kwds : `dict`
```

Comment:
I'd do It something like

```
kwds
   Additional keyword arguments used to construct or augment the `DataId`. See `TODO` parameters.
```

Then make `TODO` the API reference to whatever the `self.registry.addDataset` method is. This way the user will actually be able to figure out what these keyword arguments are.